plugin/db/mysql/binlog.go

```diff
@@ -0,0 +1,265 @@
+		set, err := parseBinlogEventPayloadBlock(block, "SET")
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "failed to parse the SET payload section of the UPDATE event")
+		}
+		return where, set, nil
+	case WriteRowsEvent:
+		// Example block:
+		// ## INSERT INTO `database`.`table`
+		// ### SET
+		// ###   @1=x
+		//       ...
+		set, err := parseBinlogEventPayloadBlock(block, "SET")
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "failed to parse the SET payload section of the INSERT event")
+		}
+		return nil, set, nil
+	default:
+		return nil, nil, errors.Errorf("invalid DML binlog event type %s", t.String())
+	}
+}
+
+type binlogEvent struct {
+	Type       binlogEventType
+	Database   string
+	Table      string
+	DataBefore [][]string
+	DataAfter  [][]string
+	ThreadID   string
+}
+
+func parseBinlogEvent(binlogText string) (*binlogEvent, error) {
+	lines := strings.Split(binlogText, "\n")
+	if len(lines) < 2 {
+		return nil, errors.Errorf("invalid mysqlbinlog dump string: must be at least 2 lines")
+	}
+	if !strings.HasPrefix(lines[0], "# at") {
+		return nil, errors.Errorf("invalid mysqlbinlog dump string: must starts with \"# at\"")
+	}
+
+	// The second line must contain the event header information.
+	// E.g., "#221017 14:25:24 server id 1  end_log_pos 916 CRC32 0x896854fc 	Write_rows: table id 259 flags: STMT_END_F"
+	header := lines[1]
+
+	var rowEvent *binlogEvent
+	var err error
+	body := lines[2:]
+	if strings.Contains(header, "Delete_rows") {
+		rowEvent, err = parseBinlogEventDML(DeleteRowsEvent, body)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to parse DELETE event")
+		}
+	} else if strings.Contains(header, "Update_rows") {
+		rowEvent, err = parseBinlogEventDML(UpdateRowsEvent, body)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to parse UPDATE event")
+		}
+	} else if strings.Contains(header, "Write_rows") {
+		rowEvent, err = parseBinlogEventDML(WriteRowsEvent, body)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to parse INSERT event")
+		}
+	} else if strings.Contains(header, "Query") {
+		rowEvent, err = parseBinlogEventQuery(header)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to parse QUERY event")
+		}
+	} else {
+		// The binlog event type is none of DELETE/UPDATE/INSERT. Skip for now.
+		rowEvent = nil
+	}
+
+	return rowEvent, nil
+}
+
+func parseBinlogEventQuery(header string) (*binlogEvent, error) {
+	matches := regexpThreadID.FindStringSubmatch(header)
+	if len(matches) != 2 {
+		return nil, errors.Errorf("failed to parse thread ID from the QUERY event header: %q", header)
+	}
+	return &binlogEvent{
+		Type:     QueryEvent,
+		ThreadID: matches[1],
+	}, nil
+}
+
+func parseBinlogEventDML(eventType binlogEventType, body []string) (*binlogEvent, error) {
+	if len(body) < eventType.MinBlockLen() {
+		return nil, errors.Errorf("invalid %s event body, must be at least %d lines", eventType.String(), eventType.MinBlockLen())
```

Comment:
include the similar but got %q as in l206

Reply:
SG.